### PR TITLE
[BEAM-3418] Send worker_id in all grpc channels to runner harness

### PR DESCRIPTION
**Pull request by angoenka**
DESCRIPTION HERE

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by angoenka** [comment]
@aaltay @lukecwik @robertwb Can you please have a look!

File: sdks/python/setup.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def get_version():
     'avro>=1.8.1,<2.0.0',
     'crcmod>=1.7,<2.0',
     'dill==0.2.6',
-    'grpcio>=1.0,<2',
```
Comment (aaltay) [inline sdks/python/setup.py]: What is the reason for this change? AFAIK, some runner (Dataflow) requires grpcio 1.3? Also, it is became more restrictive for users each time we reduce the list of allowed versions for a dependency.

Comment (lukecwik) [reply sdks/python/setup.py]: grpcio 1.8 is the first python version which allows passing/receiving client headers

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    collections.namedtuple('_ClientCallDetails',
+                           ('method', 'timeout', 'metadata', 'credentials')),
+    grpc.ClientCallDetails):
+  pass
+
+
+class WorkerIdInterceptor(grpc.StreamStreamClientInterceptor):
+
+  # Unique worker Id for this worker.
+  _worker_id = os.environ['WORKER_ID'] if os.environ.has_key(
```
Comment (herohde) [inline sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: The ID I think you want is this one: https://github.com/apache/beam/blob/master/sdks/python/container/boot.go#L41

Comment (angoenka) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Yes, For testing I am modifying the internal version of boot.go.
I was using woker_id to be more explicit but we can change it to just id.

File: sdks/python/setup.py (hunk 1 above)
Comment (aaltay) [reply sdks/python/setup.py]: Sounds good. Just try running dataflow runner test before finishing this PR.

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py (hunk 2 above)
Comment (herohde) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Sorry if I was unclear: this PR should also change the referenced boot.go code to set the WORKER_ID. The code here should also fail if that ID is not found -- the ID is set by the runner so that the gRPC servers can know who's who; they are not just unique IDs.

Comment (angoenka) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Added the relevant boot.go changes

File: sdks/python/setup.py (hunk 1 above)
Comment (angoenka) [reply sdks/python/setup.py]: Sure, Tried running apache_beam.runners.dataflow.dataflow_runner_test.DataflowRunnerTest

Comment (aaltay) [reply sdks/python/setup.py]: I meant running a job on dataflow using: https://github.com/apache/beam/blob/master/.test-infra/jenkins/job_beam_PostCommit_Python_Verify.groovy

Commenting "Run Python PostCommit" on the PR should trigger one.

**Comment by angoenka** [comment]
@herohde Please have a look. I have updated the PR to include the boot.go changes. However I am not sure how the actual id is passed to the boot.go as a flag. I made similar change in internal version of boot.go and using the new worker image.

**Comment by angoenka** [comment]
retest this please

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 3]
```diff
@@ -176,6 +176,13 @@ def __init__(self, packages, options, environment_version, pipeline_url):
     if self.debug_options.experiments:
       for experiment in self.debug_options.experiments:
         self.proto.experiments.append(experiment)
+    # Add MULTIPLE_SDK_CONTAINERS flag if its not already present. Do not add
+    # the flag if 'NO_MULTIPLE_SDK_CONTAINERS' is present.
+    # TODO: Cleanup MULTIPLE_SDK_CONTAINERS once we depricate Python SDK till
+    # version 2.4.
+    if ('MULTIPLE_SDK_CONTAINERS' not in self.proto.experiments and
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: How confident are we to make this a default behavior for 2.5?

Comment (angoenka) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: I expect this CL to get in 2.5.
In a way this flag is required to help router distinguish between old SDK (sdk till 2.4) and new SDK (sdk from 2.5). So once we do not have any sdk which is older than 2.5, we don't need to distinguish between sdk atleast for MultiSdk functionality and hence it automatically becomes the default feature.

Comment (angoenka) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: @robertwb @aaltay I am planning to make this feature opt out for new SDKs. Instead should we keep it opt in?

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 4]
```diff
@@ -176,6 +176,13 @@ def __init__(self, packages, options, environment_version, pipeline_url):
     if self.debug_options.experiments:
       for experiment in self.debug_options.experiments:
         self.proto.experiments.append(experiment)
+    # Add MULTIPLE_SDK_CONTAINERS flag if its not already present. Do not add
+    # the flag if 'NO_MULTIPLE_SDK_CONTAINERS' is present.
+    # TODO: Cleanup MULTIPLE_SDK_CONTAINERS once we depricate Python SDK till
```
Comment (lukecwik) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: `depricate` -> `deprecate`

**Comment by angoenka** [comment]
@aaltay Can you please take a look?

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py (hunk 2 above)
Comment (aaltay) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Do we want to fail if WORKER_ID is not found?

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Client Interceptor to inject worker_id"""
+from __future__ import absolute_import
+from __future__ import division
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Do we need `print_function` and `division` imports?

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  def __init__(self):
+    pass
+
+  def intercept_stream_stream(self, continuation, client_call_details,
+                              request_iterator):
+    metadata = []
+    if client_call_details.metadata is not None:
+      metadata = list(client_call_details.metadata)
+    metadata.append(('worker_id', self._worker_id))
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Would it be an error (or expected) for client_call_details to already have worker_id?

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 7]
```diff
@@ -43,6 +44,8 @@ def __init__(self, control_address, worker_count):
     self._worker_count = worker_count
     self._worker_index = 0
     self._control_channel = grpc.insecure_channel(control_address)
+    self._control_channel = grpc.intercept_channel(self._control_channel,
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py]: Should we simplify this as:

`self._control_channel = grpc.intercept_channel(grpc.insecure_channel(control_address), WorkerIdInterceptor())`

File: sdks/python/apache_beam/runners/worker/log_handler.py [hunk 8]
```diff
@@ -49,6 +50,8 @@ class FnApiLogRecordHandler(logging.Handler):
   def __init__(self, log_service_descriptor):
     super(FnApiLogRecordHandler, self).__init__()
     self._log_channel = grpc.insecure_channel(log_service_descriptor.url)
+    self._log_channel = grpc.intercept_channel(self._log_channel,
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/worker/log_handler.py]: (Same simplification comment applies here.)

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 9]
```diff
@@ -177,6 +177,14 @@ def __init__(self, packages, options, environment_version, pipeline_url):
...
       for experiment in self.debug_options.experiments:
         self.proto.experiments.append(experiment)
+    # Add use_multiple_sdk_containers flag if its not already present. Do not
+    # add the flag if 'no_use_multiple_sdk_containers' is present.
+    # TODO: Cleanup use_multiple_sdk_containers once we deprecate Python SDK
+    # till version 2.4.
+    if (job_type.startswith('FNAPI_') and
+        'use_multiple_sdk_containers' not in self.proto.experiments and
+        'no_use_multiple_sdk_containers' not in self.proto.experiments):
+      self.proto.experiments.append('use_multiple_sdk_containers')
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: It is preferable to modify debug_options.experiments (as done above for `runner_harness_override`). This also properly helps with updating the user visible pipeline options in the UI, and it will auto added to the proto by the loop above.

It would also help combine things related to `if job_type.startswith('FNAPI_'):` in a single place.

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py (hunk 5 above)
Comment (angoenka) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: No, we don't need these imports. 
I added them based to resolve the compatibility issue between python 2 and 3 based on https://docs.python.org/3/howto/pyporting.html

Should I remove them?

Comment (aaltay) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Let's remove them, if they are not needed now. I do not see print() or / being used here.

File: sdks/python/apache_beam/runners/worker/worker_id_interceptor.py (hunk 2 above)
Comment (aaltay) [reply sdks/python/apache_beam/runners/worker/worker_id_interceptor.py]: Do we really need to be backward compatible? This is mostly new code with no production usage. I would prefer to not have it succeed like this. But if you think this is necessary in the interim, we can add a TODO to remove the UUID generation.

**Comment by aaltay** [comment]
Could you squash your commits?

**Comment by angoenka** [comment]
Sure, I will squash them.
